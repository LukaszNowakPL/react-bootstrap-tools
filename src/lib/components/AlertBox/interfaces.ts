export interface AlertBoxProps {
  id?: string;
  dismissible?: boolean;
  closeLabel?: string;
  variant?: VariantValues;
  withIcon?: string | true;
  message: any;
  waitingItems?: number;
  className?: string;
  iconClassName?: string;
  withCollapse?: true;
  onEnter?: (id: string) => void;
  onEntering?: (id: string) => void;
  onEntered?: (id: string) => void;
  onExit?: (id: string) => void;
  onExiting?: (id: string) => void;
  onExited?: (id: string) => void;
  hideAfter?: number;
}

export interface AlertBoxState {
  in: boolean;
  shouldOpen: boolean;
  shouldClose: boolean;
  id: string;
}

export interface GetAlertProps {
  props: AlertBoxProps;
  state: AlertBoxState;
  methods: GetAlertPropsMethods;
}

export interface GetAlertPropsMethods {
  onClose: () => void;
}

export interface AlertProps {
  id: string;
  dismissible?: boolean;
  closeLabel?: string;
  variant?: VariantValues;
  className?: string;
  iconClassName?: string;
  onClose?: () => void;
}

export interface GetTransitionProps {
  state: AlertBoxState;
  methods: GetTransitionPropsMethods;
}

export interface GetTransitionPropsMethods {
  onEnter: () => void;
  onEntering: () => void;
  onEntered: () => void;
  onExit: () => void;
  onExiting: () => void;
  onExited: () => void;
}

export interface TransitionProps {
  in: boolean;
  onEnter: () => void;
  onEntering: () => void;
  onEntered: () => void;
  onExit: () => void;
  onExiting: () => void;
  onExited: () => void;
  timeout?: number;
}

export interface ConfigInterface {
  defaultIcon: string;
  variantIcons: object;
  defaultWrapperClassName: string;
  defaultIconClassName: string;
}

type VariantValues =
  | "primary"
  | "secondary"
  | "success"
  | "danger"
  | "warning"
  | "info"
  | "dark"
  | "light";
