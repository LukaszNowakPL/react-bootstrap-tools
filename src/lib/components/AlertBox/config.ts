import { ConfigInterface } from "./interfaces";

const config: ConfigInterface = {
  defaultIcon: "info_outline",
  variantIcons: {
    primary: "info_outline",
    secondary: "info_outline",
    success: "check_circle_outline",
    danger: "highlight_off",
    warning: "error_outline",
    info: "info_outline",
    dark: "info_outline",
    light: "info_outline"
  },
  defaultWrapperClassName: "AlertBox__wrapper",
  defaultIconClassName: "AlertBox__icon"
};

export default config;
