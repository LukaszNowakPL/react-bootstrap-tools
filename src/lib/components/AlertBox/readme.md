# AlertBox

Component for presenting short info.

![AlertBox example](../../../../readme_img/AlertBoxExample.png)

## Example of usage

```
import React from 'react';
import { AlertBox } from 'react-bootstrap-tools';
import "react-bootstrap-tools/build/static/css/index.css";

const SimplestExample = () => (
    <AlertBox variant='warning' message='warning message content' />
);

export default SimplestExample;
```

## Available props

- string | HtmlElement `message` : Message to show inside the box. It may be textual message as well as any html component.

## Available optional props

- string `id` : Prop to identify inside group of same components.
- boolean `dismissible` : Flag if component should contain `close` button.
  Additional props connected with `dismissible` prop:
  - string `closeLabel` : Label shown as a `title` during hovering on `close` button
- string `variant` : Bootstrap's variant property of an alert.<br />
  Available options:
  - primary
  - secondary
  - success
  - danger
  - warning
  - info
  - dark
  - light
- string | true `withIcon` : Flag if component should contain an additional material icon to decorate the message.
  Available values:
  - boolean : If set `true`, type of material icon will be computed using `variant` value
  - string : Material icon id. Click [here](https://material.io/tools/icons/?style=baseline) to check available icons library.
    Additional props connected with `withIcon` prop:
  - string `iconClassName` : Additional class name to custom your styling of provided icon
- number `waitingItems` : If on use, component will be decorated with a `Badge` component to present the prop's value. This prop is useful with multiple information popped at the same time and therefore bundled into a group in order to not to cover an application UI. You can easy bundle components up using `AlertBoxGroup` component.
- string `className` : Additional class name to custom your styling of a component.
- boolean `withCollapse` : Flag if component should be wrapped with Bootstrap's `Collapse` component. If not set (or false) component is wrapped using Bootstrap's `Fade` component.
- number `hideAfter` : Number of milliseconds the component should start hiding after it's been fully entered.
- callback `onEnter` : Callback function fired when component started to show. It passes string `id` arg.
- callback `onEntering` : Callback function fired when component is showing up. It passes string `id` arg.
- callback `onEntered` : Callback function fired when component has fully shown up. It passes string `id` arg.
- callback `onExit` : Callback function fired when component started to hide. It passes string `id` arg.
- callback `onExiting` : Callback function fired when component is hiding. It passes string `id` arg.
- callback `onExited` : Callback function fired when component is fully hidden. It passes string `id` arg.

## Styling

The component exposes `AlertBox__wrapper` class attached to it's wrapper. Use example code below to manipulate component's position to suit your needs

```
.AlertBox__wrapper {
  position: absolute;
  width: 100%;
  padding: 4px 10%;
  top: 25px;
}
```

The component exposes `AlertBox__icon` and `AlertBox__badge` classes to manipulate icon and badge styling.
To avoid itches on vertical-align between elements and text message use those classes together with shipped stylesheet.

```
import "react-bootstrap-tools/build/static/css/index.css";
```

If you use component with `withIcon` prop don't forget to attach Material icon stylings

```
<link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
```