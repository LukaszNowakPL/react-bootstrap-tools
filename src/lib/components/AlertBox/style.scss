.AlertBox {
  &__wrapper {
  }

  &__icon {
    margin-right: 5px;
    vertical-align: text-bottom;
  }

  &__badge {
    margin-right: 5px;
    vertical-align: text-bottom;
  }
}
