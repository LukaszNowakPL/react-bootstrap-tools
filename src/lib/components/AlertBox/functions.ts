import {
  AlertBoxProps,
  AlertProps,
  TransitionProps,
  GetTransitionProps,
  GetAlertProps
} from "./interfaces";
import config from "./config";
import { GlyphiconProps } from "../Glyphicon/interfaces";

/**
 * Returns props used by Alert component
 * @param props
 */
export const getAlertProps = (conf: GetAlertProps) => {
  const { props, state, methods } = conf;
  const { dismissible, closeLabel, variant, className } = props;
  const { id } = state;

  const alertProps: AlertProps = { id };

  if (dismissible) {
    alertProps.dismissible = dismissible;
    alertProps.onClose = methods.onClose;
    if (closeLabel) alertProps.closeLabel = closeLabel;
  }
  if (variant) alertProps.variant = variant;
  if (className) alertProps.className = className;

  return alertProps;
};

/**
 * Returns props used by transition component
 * @param props
 */
export const getTransitionProps = (conf: GetTransitionProps) => {
  const { state, methods } = conf;

  const transitionProps: TransitionProps = {
    in: state.in,
    onEnter: methods.onEnter,
    onEntering: methods.onEntering,
    onEntered: methods.onEntered,
    onExit: methods.onExit,
    onExiting: methods.onExiting,
    onExited: methods.onExited
  };

  return transitionProps;
};

/**
 * Returns icon depending on variant type of Alert
 * @param variant
 */
export const getCustomIconType = variant => {
  return config.variantIcons[variant] || config.defaultIcon;
};

/**
 * Returns props used by Glyphicon component
 * @param props
 */
export const getGlyphiconProps = (props: AlertBoxProps) => {
  const { withIcon, variant, iconClassName } = props;
  const icon = withIcon === true ? getCustomIconType(variant) : withIcon;

  const glyphiconProps: GlyphiconProps = {
    icon
  };

  if (iconClassName) {
    glyphiconProps.className = config.defaultIconClassName.concat(
      " " + iconClassName
    );
  } else {
    glyphiconProps.className = config.defaultIconClassName;
  }

  return glyphiconProps;
};
