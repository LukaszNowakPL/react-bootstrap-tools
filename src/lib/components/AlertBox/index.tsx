import * as React from "react";
import { PureComponent } from "react";
import {
  AlertBoxProps,
  AlertProps,
  TransitionProps,
  AlertBoxState,
  GetTransitionPropsMethods
} from "./interfaces";
import { Alert, Badge, Collapse, Fade } from "react-bootstrap";
import Glyphicon from "../Glyphicon";
import { GlyphiconProps } from "../Glyphicon/interfaces";
import {
  getAlertProps,
  getGlyphiconProps,
  getTransitionProps
} from "./functions";
import "./style.scss";

const IconComponent = (props: AlertBoxProps) => {
  const { withIcon } = props;

  if (withIcon) {
    const glyphiconProps: GlyphiconProps = getGlyphiconProps(props);

    return <Glyphicon {...glyphiconProps} />;
  }
};

const BadgeComponent = (props: AlertBoxProps) => {
  const { variant, waitingItems } = props;

  if (waitingItems) {
    return (
      <Badge variant={variant} className="AlertBox__badge">
        {waitingItems}
      </Badge>
    );
  }
};

class AlertBox extends PureComponent<AlertBoxProps, AlertBoxState> {
  displayName: string = "AlertBox";

  constructor(props) {
    super(props);

    const id = !props.id ? String(new Date().getTime()) : props.id;

    this.state = {
      shouldOpen: false,
      shouldClose: false,
      in: false,
      id
    };
  }

  componentDidMount() {
    this.setState({ shouldOpen: true });
  }

  static getDerivedStateFromProps(nextProps, nextState) {
    const { shouldOpen, shouldClose } = nextState;
    if (shouldOpen === true)
      return { ...nextState, in: true, shouldOpen: false };
    if (shouldClose === true)
      return { ...nextState, in: false, shouldClose: true };

    return { ...nextState };
  }

  onClose = () => {
    this.setState({ shouldClose: true });
  };

  onEnter = () => {
    const { onEnter } = this.props;
    onEnter && onEnter(this.state.id);
  };

  onEntering = () => {
    const { onEntering } = this.props;
    onEntering && onEntering(this.state.id);
  };

  onEntered = () => {
    const { onEntered, hideAfter } = this.props;
    onEntered && onEntered(this.state.id);

    if (hideAfter) setTimeout(this.onClose, hideAfter);
  };

  onExit = () => {
    const { onExit } = this.props;
    onExit && onExit(this.state.id);
  };

  onExiting = () => {
    const { onExiting } = this.props;
    onExiting && onExiting(this.state.id);
  };

  onExited = () => {
    const { onExited } = this.props;
    onExited && onExited(this.state.id);
  };

  render() {
    const { withIcon, waitingItems, message, withCollapse } = this.props;

    const iconComponent = withIcon ? IconComponent(this.props) : null;
    const badgeComponent = waitingItems ? BadgeComponent(this.props) : null;

    const alertProps: AlertProps = getAlertProps({
      props: this.props,
      state: this.state,
      methods: { onClose: this.onClose }
    });

    const alert = (
      <div className="AlertBox__wrapper">
        <Alert {...alertProps}>
          {iconComponent}
          {badgeComponent}
          {message}
        </Alert>
      </div>
    );

    const transitionMethods: GetTransitionPropsMethods = {
      onEnter: this.onEnter,
      onEntering: this.onEntering,
      onEntered: this.onEntered,
      onExit: this.onExit,
      onExiting: this.onExiting,
      onExited: this.onExited
    };
    const transitionProps: TransitionProps = getTransitionProps({
      state: this.state,
      methods: transitionMethods
    });

    return withCollapse ? (
      <Collapse {...transitionProps}>{alert}</Collapse>
    ) : (
      <Fade {...transitionProps}>{alert}</Fade>
    );
  }
}

export default AlertBox;
