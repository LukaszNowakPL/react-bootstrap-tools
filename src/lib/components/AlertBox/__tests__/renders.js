import React from "react";
import { shallow } from "enzyme";
import AlertBox from "../index";

describe("<AlertBox />", () => {
  const customComponentProps = {
    variant: "danger"
  };

  it("renders", () => {
    const component = shallow(<AlertBox {...customComponentProps} />);
    expect(component).toMatchSnapshot();
  });
});
