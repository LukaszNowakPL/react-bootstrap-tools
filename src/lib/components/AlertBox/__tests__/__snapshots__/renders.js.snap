// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertBox /> renders 1`] = `
<Fade
  appear={false}
  in={true}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={300}
  unmountOnExit={false}
>
  <div
    className="AlertBox__wrapper"
  >
    <Alert
      closeLabel="Close alert"
      id="testId"
      show={true}
      transition={[Function]}
    >
      test message
    </Alert>
  </div>
</Fade>
`;

exports[`<AlertBox /> renders closed (hidden) component 1`] = `
<Fade
  appear={false}
  in={false}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={300}
  unmountOnExit={false}
>
  <div
    className="AlertBox__wrapper"
  >
    <Alert
      closeLabel="Close alert"
      dismissible={true}
      id="testId"
      onClose={[Function]}
      show={true}
      transition={[Function]}
    >
      test message
    </Alert>
  </div>
</Fade>
`;

exports[`<AlertBox /> renders with Collapse transition 1`] = `
<Collapse
  appear={false}
  dimension="height"
  getDimensionValue={[Function]}
  in={true}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={300}
  unmountOnExit={false}
>
  <div
    className="AlertBox__wrapper"
  >
    <Alert
      closeLabel="Close alert"
      id="testId"
      show={true}
      transition={[Function]}
    >
      test message
    </Alert>
  </div>
</Collapse>
`;

exports[`<AlertBox /> renders with badge component 1`] = `
<Fade
  appear={false}
  in={true}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={300}
  unmountOnExit={false}
>
  <div
    className="AlertBox__wrapper"
  >
    <Alert
      closeLabel="Close alert"
      id="testId"
      show={true}
      transition={[Function]}
    >
      <Badge
        className="AlertBox__badge"
        pill={false}
      >
        4
      </Badge>
      test message
    </Alert>
  </div>
</Fade>
`;

exports[`<AlertBox /> renders with close button and label 1`] = `
<Fade
  appear={false}
  in={true}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={300}
  unmountOnExit={false}
>
  <div
    className="AlertBox__wrapper"
  >
    <Alert
      closeLabel="test label"
      dismissible={true}
      id="testId"
      onClose={[Function]}
      show={true}
      transition={[Function]}
    >
      test message
    </Alert>
  </div>
</Fade>
`;

exports[`<AlertBox /> renders with given className 1`] = `
<Fade
  appear={false}
  in={true}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={300}
  unmountOnExit={false}
>
  <div
    className="AlertBox__wrapper"
  >
    <Alert
      className="testClass"
      closeLabel="Close alert"
      id="testId"
      show={true}
      transition={[Function]}
    >
      test message
    </Alert>
  </div>
</Fade>
`;

exports[`<AlertBox /> renders with given variant 1`] = `
<Fade
  appear={false}
  in={true}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={300}
  unmountOnExit={false}
>
  <div
    className="AlertBox__wrapper"
  >
    <Alert
      closeLabel="Close alert"
      id="testId"
      show={true}
      transition={[Function]}
      variant="warning"
    >
      test message
    </Alert>
  </div>
</Fade>
`;

exports[`<AlertBox /> renders with icon component 1`] = `
<Fade
  appear={false}
  in={true}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={300}
  unmountOnExit={false}
>
  <div
    className="AlertBox__wrapper"
  >
    <Alert
      closeLabel="Close alert"
      id="testId"
      show={true}
      transition={[Function]}
    >
      <Glyphicon
        className="AlertBox__icon"
        icon="testIcon"
      />
      test message
    </Alert>
  </div>
</Fade>
`;
